### PR TITLE
Add ruff linter to pre-commit

### PR DESCRIPTION
**Pull request by Czaki**
# Description

Add ruff linter to pre-commit. Ruff is Rust written linter for python as a much faster replacement for flake8. Currently is not ready for total replacement, but maybe a nice addition as do not impact significantly on the pre-commit time run. 

Also, this PR contains multiple fixes in napari files. Mostly add missed `__all__` in `__init__.py` files.

## Type of change

- [x] Refactor
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: examples/3d_kymograph_.py
Comment (Carreau) [inline examples/3d_kymograph_.py]:
```suggestion
except ModuleNotFoundError:
```

`ImportError` could mean any other issue importing, and can hide problems.

**Review by Carreau — DISMISSED** [review]
Only one comment about ImportError -> ModuleNotFoundError.

The bare except is was even worse, even ImportError is an improvement.

**Review by goanpeca — APPROVED** [review]

**Comment by brisvag** [comment]
I'm not convinced about the examples changes... Examples are supposed to be a quick way to get introduced to stuff, and I find it very distracting to see all those gibberish comments. I know what they mean, but anyone sho doesn't maintain a project probably would just be confused.

I'd say either we fix the examples so they obey the rules, or we skip examples.

**Comment by Czaki** [comment]
@Carreau You request changes, and it is blocking this PR. I think that I have addressed your worries but maybe not all?



> I'm not convinced about the examples changes... Examples are supposed to be a quick way to get introduced to stuff, and I find it very distracting to see all those gibberish comments.

But if you open this code in any Python IDE same lines without these comments will be highlighted as errors.

**Comment by Czaki** [comment]
@brisvag I have workaround `noqa` clauses

File: examples/dynamic-projections-dask.py
Comment (brisvag) [inline examples/dynamic-projections-dask.py:40]: ~You could replace the above comment with a docstring to avoid the warning.~ ignore that :P

File: examples/tiled-rendering-2d_.py
Comment (brisvag) [inline examples/tiled-rendering-2d_.py:25]: does this have to be before import of napari? Otherwise, we can move it below to avoid the import lint errors.

Comment (Czaki) [reply examples/tiled-rendering-2d_.py:25]: Yes. It needs to be before. Currently, the octree is resolved on startup.

It is the weak point of napari but out of this PR  scope.

**Review by brisvag — APPROVED** [review]

**Comment by Carreau** [comment]
> @Carreau You request changes, and it is blocking this PR. I think that I have addressed your worries but maybe not all?

Feel free to dismiss reviews – or al least you should be able to dismiss them. I was not requesting changes as a way to block, but just suggestions.

**Comment by Czaki** [comment]
> Feel free to dismiss reviews – or al least you should be able to dismiss them. I was not requesting changes as a way to block, but just suggestions.

How I could do this?

**Comment by brisvag** [comment]
image

**Comment by Czaki** [comment]
If I had known this earlier... 

Thanks @brisvag and @Carreau

**Comment by brisvag** [comment]
> If I had known this earlier...

Oh no, what have I done! I created a monster :fearful:

**Pull request merged**
